File: swift/lib/Sources/Data/MappingList.swift
import Foundation

/// The shared state is guarded by the lock below, which is what the unchecked conformance asserts.
public class MappingList: @unchecked Sendable {
    public static let shared: MappingList = .init()
    private init() {}

    private let lock = NSLock()
    private var list = [String: MappingItem]()

    public func add(name: String, item: MappingItem) {
        lock.lock()
        defer { lock.unlock() }

        list[name] = item
    }

    public func find(_ name: String) -> MappingItem? {
        lock.lock()
        defer { lock.unlock() }

        return list[name]
    }

    public func clear() {
        lock.lock()
        defer { lock.unlock() }

        list.removeAll()
    }

    public func count() -> Int {
        lock.lock()
        defer { lock.unlock() }

        return list.count
    }

    public func has(_ name: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        return list[name] != nil
    }
}

File: wasm/sample/src/xplpc/data/mapping-list.ts
import { MappingItem } from "../map/mapping-item";

export class MappingList {
    private static instance: MappingList;
    private list = new Map<string, MappingItem>();

    private constructor() {}

    public static shared(): MappingList {
        if (!MappingList.instance) {
            MappingList.instance = new MappingList();
        }

        return MappingList.instance;
    }

    public add(name: string, item: MappingItem): void {
        this.list.set(name, item);
    }

    public find(name: string): MappingItem | undefined {
        return this.list.get(name);
    }

    public has(name: string): boolean {
        return this.list.has(name);
    }

    public clear(): void {
        this.list.clear();
    }

    public count(): number {
        return this.list.size;
    }
}
